internal/core/items.rs

```diff
@@ -186,6 +199,283 @@ impl ItemRc {
+        &crate::component::ComponentRc,
+        Pin<ItemRef>,
+        PostVisitState,
+        crate::item_tree::VisitChildrenResult,
+    ) -> crate::item_tree::VisitChildrenResult,
+    visit_result: crate::item_tree::VisitChildrenResult,
+) -> crate::item_tree::VisitChildrenResult {
+    if let Some(item_rc) = frame.item.upgrade() {
+        let component_rc = item_rc.component();
+        let comp_ref_pin = vtable::VRc::borrow_pin(&component_rc);
+        let item_ref = comp_ref_pin.as_ref().get_item_ref(item_rc.index());
+        post_visitor(&item_rc.component(), item_ref, frame.post_visit.take().unwrap(), visit_result)
+    } else {
+        panic!("Could not upgrade item!");
+    }
+}
+
+fn unwind_last_frame<State, PostVisitState>(
+    stack: &mut Vec<Frame<State, PostVisitState>>,
+    post_visitor: &mut impl FnMut(
+        &crate::component::ComponentRc,
+        Pin<ItemRef>,
+        PostVisitState,
+        crate::item_tree::VisitChildrenResult,
+    ) -> crate::item_tree::VisitChildrenResult,
+    visit_result: crate::item_tree::VisitChildrenResult,
+) -> crate::item_tree::VisitChildrenResult {
+    if let Some(mut f) = stack.pop() {
+        unwind_frame(&mut f, post_visitor, visit_result)
+    } else {
+        crate::item_tree::VisitChildrenResult::CONTINUE
+    }
+}
+
+fn unwind_stack<State, PostVisitState>(
+    stack: &mut Vec<Frame<State, PostVisitState>>,
+    post_visitor: &mut impl FnMut(
+        &crate::component::ComponentRc,
+        Pin<ItemRef>,
+        PostVisitState,
+        crate::item_tree::VisitChildrenResult,
+    ) -> crate::item_tree::VisitChildrenResult,
+    visit_result: crate::item_tree::VisitChildrenResult,
+) -> crate::item_tree::VisitChildrenResult {
+    let mut visit_result = visit_result;
+    if !stack.is_empty() {
+        visit_result = unwind_last_frame(stack, post_visitor, visit_result);
+    }
+    visit_result
+}
+
+fn current_state_from_stack<'a, State, PostVisitState>(
+    stack: &'a [Frame<State, PostVisitState>],
+    fallback: &'a State,
+) -> &'a State {
+    if let Some(f) = stack.last() {
+        if let Some(s) = f.state.as_ref() {
+            return s;
+        }
+    }
+    fallback
+}
+
+fn visit_internal<State, PostVisitState>(
+    item_rc: &ItemRc,
+    order: crate::item_tree::TraversalOrder,
+    visitor: &mut impl FnMut(
+        &crate::component::ComponentRc,
+        Pin<ItemRef>,
+        usize,
+        &State,
+    ) -> (crate::item_tree::ItemVisitorResult<State>, PostVisitState),
+    post_visitor: &mut impl FnMut(
+        &crate::component::ComponentRc,
+        Pin<ItemRef>,
+        PostVisitState,
+        crate::item_tree::VisitChildrenResult,
+    ) -> crate::item_tree::VisitChildrenResult,
+    state: &State,
+) -> crate::item_tree::VisitChildrenResult {
+    println!("Visiting node @{} ({:?}).", item_rc.index(), order);
+    let mut current_item = item_rc.clone();
+    let mut state_stack: Vec<Frame<State, PostVisitState>> = Vec::new();
```

Comment:
> I don't know if there is an intrinsic difference in performance.

I do not expect any, but I did not measure. My main concern is keeping the complex logic of traversing through Repeaters concentrated in one place.

> Would be nice if it would not allocates. Or at least have a good justification for it. If it does perform better, or does it allow to reduce our stack space? (IIRC the tree visit was actually not involve in the stack peak use)

I am with you there. I have been thinking how to best avoid keeping a stack, but I could not come up with a way to avoid that for the generic case of having PostVisitState:-(

Right now we store the PostVisitState in the Stack. So if that state gets big, so will the stack usage. That could be a problem at some point, even though the PostVisitState I saw so far does not seem to be overly big.

Do you know of a good tool to measure maximum stack usage?

> I think we said we wanted to expose the array (without the offsets) to the runtime so the runtime can do the item traversal with minimal calling of function in the vtable.

Yeap, that would make a lot of sense IMHO.

> In the mean time, if this helps getting the tab focus faster, go for it.

I hope this will help to avoid the entire "traverse through Repeaters" logic a couple of times more. That should really be in one place and properly tested instead of getting ad-hoc reimplementations.